File: tools/test-pi-address.py
#!/usr/bin/env python3
"""
The Pi's published address, and the two ways it can be silently useless.

    python3 tools/test-pi-address.py

Everything that needs the Pi - the RGB composites, the model layers, the
soundings - reaches it at one address, published by the Pi and read by the
browser. When that address is wrong, nothing says so usefully: every feature
reports "the Pi did not answer", which sounds like the Pi is off.

Twice it is not off at all, and the browser knew perfectly well why:

  - the address is plain http while the site is https, so the browser refuses
    to send the request. Nothing arrives at the Pi, so nothing on the Pi can
    notice, and the Pi looks dead while being completely healthy.
  - the address is a private one on the home network, which works from the
    sofa and from nowhere else.

Both are worth refusing to publish and both are worth naming when they
happen, so this checks the Pi side refuses them and the browser side tells
them apart.
"""

import ast
import os
import re
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PUB = os.path.join(ROOT, "pi", "publish_url.py")
HTML = os.path.join(ROOT, "index.html")

passed = failed = 0


def ok(name, cond, extra=""):
    global passed, failed
    if cond:
        passed += 1
        print(f"  ok   {name}")
    else:
        failed += 1
        print(f"  FAIL {name}" + (f"  <{extra}>" if extra else ""))


src = open(PUB, encoding="utf-8").read()
html = open(HTML, encoding="utf-8").read()

ns = {"re": re}
for node in ast.parse(src).body:
    if isinstance(node, ast.FunctionDef) and node.name == "usable_from_the_web":
        exec(ast.get_source_segment(src, node), ns)
assert "usable_from_the_web" in ns, "usable_from_the_web not found"
usable = ns["usable_from_the_web"]


print("\n1. a tunnel address is fine, which is the whole point")
for url in ("https://calm-river-1234.trycloudflare.com",
            "https://wx.example.com",
            "https://wx.example.com:8443"):
    ok(f"{url} is publishable", usable(url) is None, str(usable(url)))
# 172.16 through 172.31 is the private block. 172.32 is not, and a range
# check written as a plain string prefix would get this wrong.
ok("and a public address that merely looks private is not refused",
   usable("https://172.32.4.4") is None, str(usable("https://172.32.4.4")))


print("\n2. plain http is refused, because a browser would never send it")
for url in ("http://wx.example.com", "http://192.168.0.130:8080"):
    why = usable(url)
    ok(f"{url} is refused", why is not None)
    ok("  and the reason names https rather than being vague",
       why and "https" in why, str(why))


print("\n3. so is an address only the home network can reach")
for url in ("https://192.168.0.130:8080", "https://10.25.97.201",
            "https://172.16.4.4", "https://172.31.255.1",
            "https://gwcfc-pi.local", "https://localhost:8080",
            "https://127.0.0.1", "https://169.254.1.1"):
    why = usable(url)
    ok(f"{url} is refused", why is not None, str(why))


print("\n4. refusing means not publishing, not publishing something else")
# The pinned file is read by pinned_url, and that is where a bad address
# would get in. Publishing it would replace a working tunnel address with
# one that cannot work, and the failure would surface in a browser somewhere
# else entirely, hours later.
ok("the pinned address is checked before it is used",
   "usable_from_the_web" in src.split("def pinned_url")[1][:600],
   "pinned_url does not check")
ok("and a refused one becomes None rather than being published",
   re.search(r"def pinned_url.*?log\(f\"ignoring the pinned address.*?return None",
             src, re.S) is not None)
ok("with the file to delete named, since that is the fix",
   "remove {PINNED}" in src or "remove {PINNED} to go back" in src)
# The tunnel's own address is matched by a regex that only accepts https
# trycloudflare names, so that path was never able to publish plain http.
ok("the tunnel's own address was already https only",
   'URL_RE = re.compile(r"https://' in src)


print("\n5. the browser tells the same three cases apart")
ok("there is one place that works out why, rather than a guess per feature",
   "function _hdWhyUnreachable" in html)
for token, what in [
    ("'mixed'", "http behind an https page"),
    ("'lan'", "a private address"),
    ("'down'", "genuinely not answering"),
    ("'none'", "no address published at all"),
]:
    ok(f"it can say: {what}", token in html.split("_hdWhyUnreachable")[1][:2600],
       token)
# The private ranges have to agree with the Pi's, or the Pi refuses to
# publish something the browser would have explained, or the other way round.
seg = html.split("function _hdWhyUnreachable")[1][:2600]
for net in ("127\\.", "10\\.", "192\\.168\\.", "169\\.254\\.", "localhost"):
    ok(f"and it knows {net.replace(chr(92), '')} is private", net in seg, net)
ok("including the awkward 172.16 to 172.31 block",
   "172\\.(1[6-9]|2\\d|3[01])" in seg)

print("\n6. every feature that reports the Pi dead uses it")
# Three separate places used to print their own guess. A message that says
# "it may be off" when the browser knows it was blocked sends whoever reads
# it to look at the wrong thing entirely.
ok("the satellite composites", "_hdWhyUnreachable().text" in html)
ok("the model panel",
   "_hdWhyUnreachable().text" in html and "hd-warn" in html)
ok("and the sounding panel", "_hdWhyUnreachable()" in html
   and "Web (no Pi needed)" in html)
# The composites are the Pi's own work and have no fallback, but the ABI
# bands in the same menu are a public service. "Satellite is broken" is
# usually "the composites are broken", and saying so saves a lot of looking.
ok("the composites say what still works without the Pi",
   "do not need the Pi at all" in html)
ok("and the sounding points at the source that needs no Pi",
   "Web (no Pi needed)" in html)

print()
if failed:
    print(f"{failed} FAILED, {passed} passed")
    sys.exit(1)
print(f"all {passed} passed")

File: tools/test-pi-address.mjs
#!/usr/bin/env node
/*
 * Finding the Pi, and surviving a published address that lies.
 *
 *     npm i playwright && node tools/test-pi-address.mjs
 *
 * On 2026-08-27 the site reported the Pi unreachable for a day while the Pi
 * answered every request put to it. Nothing was broken in the way anyone
 * looked for. The chain was:
 *
 *   - the Pi's quick tunnel rotated its name, as quick tunnels do
 *   - a database rule change meant the Pi could no longer publish the new one
 *   - so the shared document went on naming an address dead for hours
 *   - and the site believed it, because believing it was the whole design
 *
 * Two defences, and this is what tests them. The Pi has a permanent hostname
 * that cannot rotate. And no candidate address is believed on its word: they
 * are asked, and the first that answers wins. An address that does not answer
 * is not an address, however officially it was published.
 *
 * The failure this guards against is silent by nature, so every check here is
 * about which address is CHOSEN, given what does and does not answer.
 */

import { readdirSync, existsSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

let chromium;
try {
  ({ chromium } = await import('playwright'));
} catch {
  console.log('playwright is not installed, skipping. npm i playwright');
  process.exit(0);
}

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..');

const LEAFLET_STUB = `(() => {
  const chain = () => new Proxy(function(){}, {
    get: (t, k) => {
      if (k === 'getCenter')  return () => ({ lat: 35.3, lng: -97.3 });
      if (k === 'getZoom')    return () => 7;
      if (k === 'hasLayer')   return () => false;
      if (k === 'getPane')    return () => document.createElement('div');
      if (k === 'createPane') return () => document.createElement('div');
      if (k === 'getBounds')  return () => ({ getWest:()=>-100, getEast:()=>-95,
        getNorth:()=>38, getSouth:()=>33, contains:()=>true, pad(){return this;},
        getCenter:()=>({ lat:35, lng:-97 }) });
      if (k === 'then') return undefined;
      return chain();
    },
    apply: () => chain(), construct: () => chain(),
  });
  Object.defineProperty(window, 'L',
    { value: chain(), writable: true, configurable: true });
})();`;

let pass = 0, fail = 0;
const ok = (name, cond, extra) => {
  if (cond) { pass++; console.log('  ok   ' + name); }
  else { fail++; console.log('  FAIL ' + name + (extra ? '  <' + extra + '>' : '')); }
};

function chromePath() {
  if (process.env.CHROME_PATH) return process.env.CHROME_PATH;
  const root = '/opt/pw-browsers';
  try {
    for (const d of readdirSync(root)) {
      if (!d.startsWith('chromium-')) continue;
      const p = join(root, d, 'chrome-linux', 'chrome');
      if (existsSync(p)) return p;
    }
  } catch { /* let Playwright try its own */ }
  return undefined;
}

const browser = await chromium.launch({ executablePath: chromePath() });
const page = await browser.newPage();
const errors = [];
page.on('pageerror', e => errors.push(e.message));

await page.addInitScript(LEAFLET_STUB);
await page.route('**://**', r =>
  r.request().url().startsWith('file://') ? r.continue() : r.abort());
await page.goto('file://' + join(ROOT, 'index.html'),
                { waitUntil: 'domcontentloaded' });
await page.waitForTimeout(4000);

// The whole world, as far as address resolution is concerned: which hosts are
// alive, what the shared document says, and what is saved in this browser.
await page.evaluate(() => {
  window.__world = { alive: [], published: null, saved: null, asked: [] };
  // Bare assignments: these are top-level lets, which are lexical bindings
  // rather than properties of window.
  _hdFetchPublished = async () => window.__world.published;
  window.__realAnswers = async (base) => {
    if (!base) return false;
    window.__world.asked.push(base);
    return window.__world.alive.includes(base);
  };
  _hdAnswers = window.__realAnswers;
  window.__resolve = async () => {
    window.__world.asked = [];
    _hdResolveCache = { at: 0, base: null, pr: null };
    try { localStorage.removeItem(HD_BASE_KEY); } catch (e) {}
    if (window.__world.saved) {
      try { localStorage.setItem(HD_BASE_KEY, window.__world.saved); } catch (e) {}
    }
    const base = await _hdResolveBase(true);
    let left = null;
    try { left = localStorage.getItem(HD_BASE_KEY); } catch (e) {}
    return { base, asked: window.__world.asked.slice(), saved: left };
  };
});

const FIXED = await page.evaluate(() => HD_FIXED_BASE);
const TUNNEL_OLD = 'https://dead-quick-tunnel.trycloudflare.com';
const TUNNEL_NEW = 'https://live-quick-tunnel.trycloudflare.com';

console.log('\n1. the permanent hostname exists and is a real https address');
ok('there is one', !!FIXED, String(FIXED));
ok('it is https, because the site is and a browser will not mix them',
   /^https:\/\//.test(FIXED || ''), String(FIXED));
ok('and it is not a quick tunnel, which is the thing that rotates',
   !/trycloudflare/.test(FIXED || ''), String(FIXED));

console.log('\n2. the ordinary day: the published address answers, and wins');
{
  const r2 = await page.evaluate(async ({ nu }) => {
    window.__world = { alive: [nu, HD_FIXED_BASE], published: nu,
                       saved: null, asked: [] };
    return window.__resolve();
  }, { nu: TUNNEL_NEW });
  ok('the published address is used', r2.base === TUNNEL_NEW, r2.base);
  ok('and the permanent name is not preferred over a deployment\'s own Pi',
     r2.base !== FIXED, r2.base);
}

console.log('\n3. the outage: the published address is dead, the fixed one is not');
{
  const r = await page.evaluate(async ({ dead }) => {
    window.__world = { alive: [HD_FIXED_BASE], published: dead,
                       saved: null, asked: [] };
    return window.__resolve();
  }, { dead: TUNNEL_OLD });
  ok('the dead published address is NOT used', r.base !== TUNNEL_OLD, r.base);
  ok('the permanent hostname is used instead', r.base === FIXED, r.base);
  ok('and the dead one was genuinely asked, not skipped on a guess',
     r.asked.includes(TUNNEL_OLD), JSON.stringify(r.asked));
}

console.log('\n4. everything is asked at once, not one timeout after another');
{
  const r = await page.evaluate(async ({ dead, mine }) => {
    window.__world = { alive: [], published: dead, saved: mine, asked: [] };
    const t0 = performance.now();
    // Each probe holds for 200ms. Three in series is 600, in parallel is 200.
    _hdAnswers = async (base) => {
      window.__world.asked.push(base);
      await new Promise(res => setTimeout(res, 200));
      return false;
    };
    const out = await window.__resolve();
    // Put the world-driven probe back. Leaving this always-false stub in place
    // made three later checks pass for the wrong reason: with nothing ever
    // answering, "the dead address was refused" is true because EVERY address
    // was refused, which proves nothing at all.
    _hdAnswers = window.__realAnswers;
    return { ms: performance.now() - t0, asked: out.asked, base: out.base };
  }, { dead: TUNNEL_OLD, mine: 'https://saved.example.com' });
  ok('all three candidates are tried', r.asked.length === 3,
     JSON.stringify(r.asked));
  ok('and in parallel: three 200ms probes take about 200ms, not 600',
     r.ms < 450, Math.round(r.ms) + 'ms');
  ok('and the world-driven probe is back for the checks below',
     await page.evaluate(() => _hdAnswers === window.__realAnswers));
}

console.log('\n5. when nothing answers it still names the address it was given');
{
  const r = await page.evaluate(async ({ dead }) => {
    window.__world = { alive: [], published: dead, saved: null, asked: [] };
    const out = await window.__resolve();
    _hdBase = out.base;
    return { base: out.base, why: _hdWhyUnreachable() };
  }, { dead: TUNNEL_OLD });
  ok('it hands back the published address rather than null',
     r.base === TUNNEL_OLD, r.base);
  ok('so the failure message can say WHICH address did not answer',
     (r.why.text || '').includes(TUNNEL_OLD), r.why.text);
}

console.log('\n6. a saved address is the last resort, and is cleared when wrong');
{
  const r = await page.evaluate(async ({ mine }) => {
    window.__world = { alive: [mine], published: null, saved: mine, asked: [] };
    return window.__resolve();
  }, { mine: 'https://my-own-pi.example.com' });
  ok('a saved address that answers is used when nothing else does',
     r.base === 'https://my-own-pi.example.com', r.base);

  const r2 = await page.evaluate(async ({ mine, nu }) => {
    window.__world = { alive: [nu, HD_FIXED_BASE], published: nu,
                       saved: mine, asked: [] };
    return window.__resolve();
  }, { mine: 'https://stale.example.com', nu: TUNNEL_NEW });
  ok('a stale saved address does not shadow a live published one',
     r2.base === TUNNEL_NEW, r2.base);
  ok('and it is deleted, so it cannot come back later', r2.saved === null,
     String(r2.saved));
}

console.log('\n7. the watcher cannot drag a working site onto a dead address');
{
  const r = await page.evaluate(async ({ dead, mine }) => {
    _hdBase = mine;
    _hdWatchBusy = false;
    window.__world = { alive: [mine], published: dead, saved: null, asked: [] };
    let applied = null;
    const realApply = _hdApplyNewBase;
    _hdApplyNewBase = (b) => { applied = b; return true; };
    await _hdPollAddress();
    _hdApplyNewBase = realApply;
    return { applied, base: _hdBase, asked: window.__world.asked };
  }, { dead: TUNNEL_OLD, mine: 'https://working.example.com' });
  ok('a newly published address that does not answer is refused',
     r.applied === null, String(r.applied));
  ok('and it was asked before being refused',
     r.asked.includes(TUNNEL_OLD), JSON.stringify(r.asked));
}

console.log('\n8. and it still follows a tunnel that really did move');
{
  const r = await page.evaluate(async ({ nu, mine }) => {
    _hdBase = mine;
    _hdWatchBusy = false;
    window.__world = { alive: [nu], published: nu, saved: null, asked: [] };
    let applied = null;
    const realApply = _hdApplyNewBase;
    _hdApplyNewBase = (b) => { applied = b; return true; };
    await _hdPollAddress();
    _hdApplyNewBase = realApply;
    return { applied };
  }, { nu: TUNNEL_NEW, mine: 'https://working.example.com' });
  ok('a live new address is followed, which is what the watcher is for',
     r.applied === TUNNEL_NEW, String(r.applied));
}

console.log('\n9. an address that dies WITHOUT the document changing is noticed');
{
  // The exact shape of the outage: the published address never changed, so the
  // watcher had nothing to react to, while the address everyone was on was
  // already dead.
  const r = await page.evaluate(async ({ dead }) => {
    _hdBase = dead;
    _hdHealthAt = 0;
    window.__world = { alive: [HD_FIXED_BASE], published: dead,
                       saved: null, asked: [] };
    _hdResolveCache = { at: 0, base: null, pr: null };
    let applied = null;
    const realApply = _hdApplyNewBase;
    _hdApplyNewBase = (b) => { applied = b; return true; };
    await _hdCheckCurrent();
    _hdApplyNewBase = realApply;
    return { applied, asked: window.__world.asked };
  }, { dead: TUNNEL_OLD });
  ok('the address in use is checked, not only the published one',
     r.asked.includes(TUNNEL_OLD), JSON.stringify(r.asked));
  ok('and when it has died the site moves itself to one that works',
     r.applied === FIXED, String(r.applied));
}

console.log('\n10. resolving is not repeated dozens of times a page load');
{
  const r = await page.evaluate(async ({ nu }) => {
    window.__world = { alive: [nu], published: nu, saved: null, asked: [] };
    _hdResolveCache = { at: 0, base: null, pr: null };
    const first = await _hdResolveBase(true);
    const n = window.__world.asked.length;
    // Twenty places ask for the address on a page load.
    for (let i = 0; i < 20; i++) await _hdResolveBase();
    return { first, probesForOne: n, total: window.__world.asked.length };
  }, { nu: TUNNEL_NEW });
  ok('twenty callers cost one resolve, not twenty',
     r.total === r.probesForOne,
     `${r.total} probes for 21 calls, ${r.probesForOne} for the first`);
}

console.log('\n11. nothing threw along the way');
ok('no uncaught errors at all', errors.length === 0, errors.join(' | '));

await browser.close();
console.log(fail ? `\n${fail} FAILED, ${pass} passed` : `\nall ${pass} passed`);
process.exit(fail ? 1 : 0);
